src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,146 @@
+            //  at Microsoft.Spark.Sql.PicklingUdfWrapper`2.Execute(Int32 splitIndex, Object[] input, Int32[] argOffsets) in Microsoft.Spark\Sql\PicklingUdfWrapper.cs:line 44
+            //  at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.SingleCommandRunner.Run(Int32 splitId, Object input) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 239
+            //  at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.ExecuteCore(Stream inputStream, Stream outputStream, SqlCommand[] commands) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 139
+            Func<Column, Column> udfInt = Udf<int[], string>(array => string.Join(',', array));
+            Assert.Throws<Exception>(() => _df.Select(udfInt(_df["ages"])).Show());
+
+            // Currently, there is a workaround to support array type using ArrayList. See the example below.
+            Func<Column, Column> workingUdf = Udf<ArrayList, string>(array => string.Join(',', array.ToArray()));
+
+            Row[] rows = _df.Select(workingUdf(_df["ages"])).Collect().ToArray();
+            Assert.Equal(3, rows.Length);
+            
+            var expected = new[] { "19", "19,30", "30,40" };
+            for (int i = 0; i < rows.Length; ++i)
+            {
+                Row row = rows[i];
+                Assert.Equal(1, row.Size());
+                Assert.Equal(expected[i], row.GetAs<string>(0));
+            }
+        }
+
+        /// <summary>
+        /// UDF that returns Array type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnTypeAsArray()
+        {
+            // System.NotImplementedException: The method or operation is not implemented.
+            // at Microsoft.Spark.Sql.Row.Convert() in Microsoft.Spark\Sql\Row.cs:line 169
+            // at Microsoft.Spark.Sql.Row..ctor(Object[] values, StructType schema) in Microsoft.Spark\Sql\Row.cs:line 34
+            // at Microsoft.Spark.Sql.RowConstructor.GetRow() in Microsoft.Spark\Sql\RowConstructor.cs:line 113
+            // at Microsoft.Spark.Sql.RowCollector.Collect(ISocketWrapper socket) + MoveNext() in Microsoft.Spark\Sql\RowCollector.cs:line 36
+            // at Microsoft.Spark.Sql.DataFrame.GetRows(String funcName) + MoveNext() in Microsoft.Spark\Sql\DataFrame.cs:line 891
+            Func<Column, Column> udf = Udf<string, string[]>(
+                str => new string[] { str, str + str });
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+        }
+
+        /// <summary>
+        /// UDF that takes in Map type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithMapType()
+        {
+            // System.NotImplementedException: The method or operation is not implemented.

```

Comment:
Does the `Show()` work fine?

Reply:
Nope, `Show()` doesn't work here. Should I add an exception message with `Show()` here?

Reply:
Got this exception `Unable to cast object of type 'Microsoft.Spark.Sql.Row' to type 'System.Collections.Generic.IDictionary 2[System.String,System.String]'.`. Does this mean my input is not a MapType?

Reply:
I have added the an exception message here and shown a working example with Hashtable.